### PR TITLE
Application: opts arg to start(); app 'config' event, all deepmerged to rc config

### PR DESCRIPTION
**Pull request by loppear**
`application.start({worker_queue: {redis_url: ...}})` or `application.on('config', () => {return {worker_queue: {redis_url: ...}} })` both of those will be deepmerged, start opts take precedence.

### CODE REVIEW

File: src/Application.js [hunk 1]
```diff
@@ -116,10 +117,17 @@ export default class Application extends Dispatcher {
   _setupConfig() {
     this.setUserConfig(null, _userConfig)
 
-    this.config = Object.assign(this.config, _defaultConfig)
-    this.config = Object.assign(this.config, new ConfigurationManager(this.config).getConfig())
-    this.config = Object.assign(this.config, this._opts)
-    if(typeof this.config.debug === 'undefined') this.config.debug = (!process.env.NODE_ENV || process.env.NODE_ENV == 'development')
+    return this.emit('config').then((config={}) => {
+      if (_.isArray(config)) {
+        config = Object.assign(...config)
```
Comment (davidkellerman) [inline src/Application.js]: I'm not clear on how `config` is getting collected, but (1) do you need a `deepExtend()` here, and (2) is is safe to overwrite `config[0]` with the `Object.assign()`?

**Comment by loppear** [comment]
Documentation updated.

File: src/Application.js [hunk 2]
```diff
@@ -175,13 +183,14 @@ export default class Application extends Dispatcher {
...
   _init() {
-    this._setupConfig()
-    this._pluginInstances = {}
-    return this._loadPlugins().then(::this._boot).then(() => {
-      if (!this._appWatcher && !this.config.script && this.config.NODE_ENV != 'production') {
-        this.log.debug('Setting up App watcher')
-        this._appWatcher = new Watcher(this, this._getWatchPaths(), 'change', this._getAppIgnorePaths())
-      }
+    return this._setupConfig().then(() => {
+      this._pluginInstances = {}
```
Comment (davidkellerman) [inline src/Application.js]: Not sure of sequencing – is it safe to defer initializing `_pluginInstances`?

File: src/Application.js (hunk 1 above)
Comment (loppear) [reply src/Application.js]: yah, this should probably deepExtend for both points.

File: src/Application.js (hunk 2 above)
Comment (loppear) [reply src/Application.js]: Yep

**Pull request merged**